api/controllers/models.py

```diff
@@ -416,3 +419,74 @@ def upload_to_s3(credentials):
             {"model_id": model.id, "s3_uri": f"s3://{bucket_name}/{s3_path}"}
         )
     )
+
+
+@bottle.get("/models/<mid:int>/deploy")
+@_auth.requires_auth
+def deploy_model_from_s3(credentials, mid):
+    # Authentication (only authenticated users can redeploy models for interaction)
+    u = UserModel()
+    user_id = credentials["id"]
+    user = u.get(user_id)
+    if not user:
+        logger.error("Invalid user detail for id (%s)" % (user_id))
+        bottle.abort(404, "User information not found")
+
+    m = ModelModel()
+    model = m.getUnpublishedModelByMid(mid)
+
+    model_owner = model.uid == user.id
+
+    if (not model.is_published) and (not model_owner):
+        bottle.abort(403, "Model is not published and user is not model owner")
+
+    if not model.deployment_status == DeploymentStatusEnum.takendownnonactive:
+        bottle.abort(
+            403, "Attempting to deploy a model not taken down due to inactivity"
+        )
+
+    model_name = model.name
+
+    t = TaskModel()
+    task = t.getBtTaskId(model.tid)
+    task_code = task.task_code
+    bucket_name = task.s3_bucket
+
+    endpoint_name = model.endpoint_name
+    s3_filename = f"{endpoint_name}.tar.gz"
+    s3_path = f"torchserve/models/{task_code}/{s3_filename}"
+
+    # Update database entry
+    session = boto3.Session(
+        aws_access_key_id=config["aws_access_key_id"],
+        aws_secret_access_key=config["aws_secret_access_key"],
+        region_name=config["aws_region"],
+    )
+
+    # send SQS message
+    logger.info(f"Send message to sqs - enqueue model {model_name} for re-deployment")
+    sqs = session.resource("sqs")
+    queue = sqs.get_queue_by_name(QueueName=config["builder_sqs_queue"])
+    queue.send_message(
+        MessageBody=ujson.dumps(
+            {"model_id": model.id, "s3_uri": f"s3://{bucket_name}/{s3_path}"}
+        )
+    )
+
+    return {"status": "success"}
+
+
+@bottle.get("/models/<mid:int>/log_interaction")
```

Comment:
The purpose of this endpoint is to be able to log when a user successfully interacts with a model (so that we can update the `last_used` field; however, I also think having a separate endpoint just to update this might be a bit too much.

The alternative was to just log a model interaction in some endpoint that is hit iff a user interacts with a model successfully (for example `/examples/evaluate`)

Reply:
Also — is it ok to consider successful model interactions (i.e when the model receives something and responds to the user successfully) to update `last_used`, or should it be whenever a user attempts to interact with a model?

Reply:
Also, not sure if it is ok to only consider successful model interactions to update `last_used`. In other words, should we update `last_used` whenever a user attempts to interact with a model, or only when the user interacts successfully with a model (the model receives the input and successfully returns a response)